File: packages/analytics/src/console-sink.ts
import type { AnalyticsEvent, AnalyticsSink } from './types.js'

export interface ConsoleSinkOptions {
  /** Injected logger (defaults to globalThis.console). */
  logger?: Pick<Console, 'log' | 'groupCollapsed' | 'groupEnd'>
  /** Consent categories this sink requires. Default: none (always allowed). */
  consentCategories?: string[]
}

/**
 * Built-in `console` sink — the dev preset's default. Prints each canonical
 * envelope so engineers can see exactly what would ship over the wire.
 */
export function createConsoleSink(
  options: ConsoleSinkOptions = {},
): AnalyticsSink {
  const logger =
    options.logger ??
    (globalThis as unknown as { console: Console }).console

  return {
    name: 'console',
    consentCategories: options.consentCategories,
    deliver(batch: AnalyticsEvent[]): void {
      for (const ev of batch) {
        const label = `[analytics] ${ev.type}${ev.event ? ` ${ev.event}` : ''}`
        if (typeof logger.groupCollapsed === 'function') {
          logger.groupCollapsed(label)
          logger.log(ev)
          logger.groupEnd?.()
        } else {
          logger.log(label, ev)
        }
      }
    },
  }
}

File: packages/flutter/lib/src/analytics/console_sink.dart
/// Built-in `console` sink — the dev preset's default. Prints each canonical
/// envelope so engineers can see exactly what would ship over the wire.
library;

import 'types.dart';

/// Logger SPI (defaults to `print`). Injectable for tests / custom routing.
typedef AnalyticsLogger = void Function(String message);

class ConsoleSink implements AnalyticsSink {
  ConsoleSink({AnalyticsLogger? logger, List<String>? consentCategories})
    : _logger = logger ?? _printLogger,
      _consentCategories = consentCategories;

  static void _printLogger(String message) {
    // ignore: avoid_print
    print(message);
  }

  final AnalyticsLogger _logger;
  final List<String>? _consentCategories;

  @override
  String get name => 'console';

  @override
  List<String>? get consentCategories => _consentCategories;

  @override
  void deliver(List<AnalyticsEvent> batch, SinkDeliverContext ctx) {
    for (final ev in batch) {
      final label =
          '[analytics] ${ev.type.wire}${ev.event != null ? ' ${ev.event}' : ''}';
      _logger('$label ${ev.toJson()}');
    }
  }

  @override
  void init(SinkInitContext ctx) {}

  @override
  void flush() {}

  @override
  void shutdown() {}
}

/// Create the built-in dev-preset console sink.
AnalyticsSink createConsoleSink({
  AnalyticsLogger? logger,
  List<String>? consentCategories,
}) => ConsoleSink(logger: logger, consentCategories: consentCategories);
